hogben/simulate.py

```diff
@@ -10,368 +10,270 @@
+            to None i.e. non-magnetic
+
+        Returns:
+            list[tuple[np.ndarray]]: simulated data for the given model in the
+            form [(q, r, dr, counts),...] with one entry per spin state
+        """
+        # Non-polarised case
+        simulation = []
+        if spin_states is None:
+            for angle, points, time in self.angle_times:
+                simulated_angle = self._run_experiment(angle, points, time)
+                simulation.append(simulated_angle)
+            return simulation
+
+        # Simulate each spin state for all conditions and then add them
+        simulated_spin_states = []
+        for spin_state in spin_states:
+            simulation = []
+            for angle, points, time in self.angle_times:
+                simulated_angle = self._run_experiment(angle, points,
+                                                       time, spin_state)
+                simulation.append(simulated_angle)
+
+            simulated_spin_states.append(simulation)
+        return [np.array(simulated_spin_states)[:,i] for i in range(4)]
+
+
+    def simulate_magnetic(self, mm: bool = True, mp: bool = True,
+                          pm: bool = True, pp: bool = True) -> tuple:
+        """Simulates an experiment of a given magnetic `sample` measured
+           over a number of angles.
+        Args:
+            pp: whether to simulate "plus plus" spin state
+            pm: whether to simulate "plus minus" spin state
+            mp: whether to simulate "minus plus" spin state
+            mm: whether to simulate "minus minus" spin state
+
+        Returns:
+            tuple: model and simulated data for the given `sample`
+        """
+        models, datasets = [], []
+        # Simulate the spin states if requested.
+        for i, spin_state in enumerate([mm, mp, pm, pp]):
+            if spin_state is True:
+                model, data = self.simulate(i)
+                models.append(model)
+                datasets.append(data)
+
+        return models, datasets
+
+    def refl1d_experiment(self, q_array: np.ndarray, spin_state: Optional[
+        int] = None) -> refl1d.experiment.Experiment:
+        """Creates a Refl1D experiment for a given `sample` and `q_array`.
+        Also calculates the resolution for refl1D to be the same as is defined
+        for refnx as it is not by default.
+
+        Args:
+            q_array: Q points to use in the experiment.
+            spin_state: spin state to simulate if given a magnetic sample.
+
+        Returns:
+            refl1d.experiment.Experiment: experiment for the given `sample`.
+
+        """
+        # Transform the resolution from refnx to Refl1D format.
+        refl1d_dq = self.dq / (100 * np.sqrt(8 * np.log(2)))
+
+        # Calculate the dq array and use it to define a QProbe.
+        dq_array = q_array * refl1d_dq
+        probe = refl1d.probe.QProbe(q_array, dq_array, intensity=self.scale,
```

Comment:
self.scale is used twice in the class, but never defined (not in the init function either).